### ISSUE TITLE
[bug] NaN values in case of 0 RMSNorm weight

### ISSUE THREAD

**Issue by sohamparikh** [ServiceNow/Fast-LLM#220]
# 🐞 Describe the Bug

The backward pass for RMSNorm can result in nan if any weight is 0. I understand that this would rarely happen in typical training runs. However, we observe that several weights (including RMSNorm) often become 0 after merging checkpoints, while the model quality itself is unaffected or even superior than the individual checkpoints. 

https://github.com/ServiceNow/Fast-LLM/blob/58b6f8a6ad919a5b48048c12066d45e1b10fbd6e/fast_llm/functional/triton/normalization.py#L101 

# 🔄 Steps to Reproduce

Start a Fast-LLM training run with any single RMSNorm weight set to 0, and RMSNorm implementation set to `triton`.
Observed logs:

```console
2025-04-01 17:18:22,486 [Rank 000] Creating model...
2025-04-01 17:18:22,508 [Rank 000]   Splitting the model into 26 stages...
2025-04-01 17:18:22,513 [Rank 000]   Total parameters: 7,885,542,400 
2025-04-01 17:18:22,514 [Rank 000] Weight buffer placement:
{0: 0, 2: 0, 4: 0, 6: 0, 8: 0, 10: 0, 12: 0, 14: 0, 16: 0, 18: 0, 20: 0, 22: 0, 24: 0, 1: 1, 3: 1, 5: 1, 7: 1, 9: 1, 11: 1, 13: 1, 15: 1, 17: 1, 19: 1, 21: 1, 23: 1, 25: 1}
2025-04-01 17:18:22,514 [Rank 000] Grad buffer placement:
{0: 0, 2: 0, 4: 0, 6: 0, 8: 0, 10: 0, 12: 0, 14: 0, 16: 0, 18: 0, 20: 0, 22: 0, 24: 0, 1: 1, 3: 1, 5: 1, 7: 1, 9: 1, 11: 1, 13: 1, 15: 1, 17: 1, 19: 1, 21: 1, 23: 1, 25: 1}
2025-04-01 17:18:22,705 [Rank 000] Command run:
/usr/local/bin/fast-llm train gpt --config <redacted>
2025-04-01 17:18:22,710 [Rank 000] 
------- fast_llm.models.gpt.config.GPTTrainerConfig --------
run:
  experiment_dir: <redacted>
model:
  base_model:
    transformer:
      normalization:
        type: rms_norm
      num_layers: 24
      hidden_size: 5120
      num_attention_heads: 32
      head_groups: 8
      add_linear_biases: false
      ffn_hidden_size: 14336
      kv_channels: 128
      rotary:
        type: default
        theta: 1000000.0
      gated: true
      activation_type: silu
      init_method_std: 0.013975424859373685
      init_method_std_qkv: 0.013975424859373685
      init_method_std_attn_proj: 0.0020171788261496965
      init_method_std_mlp_1: 0.013975424859373685
      init_method_std_mlp_2: 0.0020171788261496965
      mlp_lr_scale:
      - null
    vocab_size: 131072
    use_position_embeddings: false
    tie_word_embeddings: false
    init_method_std_embed: 0.013975424859373685
    cross_entropy_impl: fused
  multi_stage:
    num_weight_buffers: 2
    num_grad_buffers: 2
    zero_stage: 3
  distributed:
    sequence_data_parallel: 2
    world_size: 128
    rank: 0
    local_world_size: 8
    timeout: 3600.0
    training_dtype: bfloat16
pretrained:
  format: mistral
  path: <redacted>
  load_config: architecture
training:
  validation:
    interval: 1000
    iterations: 25
  logs:
    interval: 2
  checkpoint:
    interval: 1000
  export:
    format: mistral
    optimizer_state: false
    interval: 1000
  train_iters: 20000
  num_workers: 8
batch:
  micro_batch_size: 1
  depth_first_micro_batches: 12
  breadth_first_micro_batches: 1
  sequential_micro_batches: 12
  batch_size: 768
  sequence_length: 16384
  micro_sequence_length: 16384
data:
  split:
  - 0.998
  - 0.002
  - 0.0
  format: file
  path:
  - <redacted>
  fim:
    rate: 0.0
  tokenizer:
    path: null
  datasets:
    Training:
      split:
      - 0.998
      - 0.002
      - 0.0
      format: file
      path:
      - <redacted>
      fim:
        rate: 0.0
      type: legacy
    Validation:
      split:
      - 0.998
      - 0.002
      - 0.0
      format: file
      path:
      - <redacted>
      fim:
        rate: 0.0
      type: legacy
    Test:
      split:
      - 0.998
      - 0.002
      - 0.0
      format: file
      path:
      - <redacted>
      fim:
        rate: 0.0
      type: legacy
profiling:
  cuda: false
optimizer:
  learning_rate:
    base: 5.0e-05
    decay_style: linear
    decay_iterations: 20000
    warmup_iterations: 100
  weight_decay: 0.1
  beta_2: 0.95
--------------------------- end ----------------------------
2025-04-01 17:18:24,751 [Rank 000] >>> Allocating 2 weight buffers (2,560.02 MiB)
2025-04-01 17:18:24,753 [Rank 000] >>> Allocating 2 grad buffers (5,120.03 MiB)
2025-04-01 17:18:24,754 [Rank 000] >>> Allocating 4 x 26 shards (940.04 MiB)
2025-04-01 17:18:24,755 [Rank 000] Total allocated: 8,620.08 MiB
2025-04-01 17:18:24,765 [Rank 000] Preparing datasets...
2025-04-01 17:18:24,766 [Rank 000] Preparing dataset. This may take several minutes.
2025-04-01 17:19:28,126 [Rank 000] Initializing training state from pretrained checkpoint at <redacted> (resetting optimizer state)...
2025-04-01 17:19:28,129 [Rank 000] The configuration parameter `('sliding_window',)=4096` is ignored during conversion. If you intend to use it in Fast-LLM, make sure to set it explicitly in the model configuration.
2025-04-01 17:19:29,888 [Rank 000] Loading index from <redacted>/model.safetensors.index.json
2025-04-01 17:19:29,891 [Rank 000] Loading from <redacted>/model-00001-of-00004.safetensors
2025-04-01 17:20:16,756 [Rank 000] Loading from <redacted>/model-00004-of-00004.safetensors
2025-04-01 17:20:30,400 [Rank 000] Loading from <redacted>/model-00002-of-00004.safetensors
2025-04-01 17:21:18,377 [Rank 000] Loading from <redacted>/model-00003-of-00004.safetensors
2025-04-01 17:22:05,603 [Rank 000] 61,606,208 state entries loaded successfully
2025-04-01 17:22:05,821 [Rank 000] done with setup ...
2025-04-01 17:22:05,822 [Rank 000] After initial setup:  allocated 8,620.08 MiB | max allocated 9,900.08 MiB | reserved 9,908.00 MiB | max reserved 9,908.00 MiB | global max reserved 9,908.00 MiB
2025-04-01 17:22:05,824 [Rank 000] Initializing PhaseType.training data iterator from sample 0...
2025-04-01 17:22:57,165 [Rank 000] Training ...
2025-04-01 17:23:41,472 [Rank 000] PhaseType.training @ iteration      2/ 20000 | consumed samples:        1,536 | consumed tokens:       25,165,824 | batch size: 768 | step time: 20220.65 ms | throughput: 257.41 tflop/s (model) | 265.24 tflop/s (hardware) | 4861.56 tokens/s/gpu | Memory allocated 8,692.08 MiB | max allocated 46,207.17 MiB | reserved 50,212.00 MiB | max reserved 50,212.00 MiB | global max reserved 50,212.00 MiB | learning rate: 1.000e-06 | loss scale:     1 | grad norm: nan | skipped iterations:   0 | nan iterations:   0 | average step time 20220.65 ms | remaining 4 days, 16:19:33  | completion 2025-04-06 09:43:14 (0.01 %) | language model loss: nan | run: 5
2025-04-01 17:24:04,210 [Rank 000] PhaseType.training @ iteration      4/ 20000 | consumed samples:        3,072 | consumed tokens:       50,331,648 | batch size: 768 | step time: 11369.04 ms | throughput: 457.83 tflop/s (model) | 471.76 tflop/s (hardware) | 8646.64 tokens/s/gpu | Memory allocated 8,692.08 MiB | max allocated 46,207.17 MiB | reserved 50,212.00 MiB | max reserved 50,212.00 MiB | global max reserved 50,212.00 MiB | learning rate: 2.000e-06 | loss scale:     1 | grad norm: nan | skipped iterations:   0 | nan iterations:   0 | average step time 15794.85 ms | remaining 3 days, 15:43:54  | completion 2025-04-05 09:07:58 (0.02 %) | language model loss: nan | run: 5
2025-04-01 17:24:26,934 [Rank 000] PhaseType.training @ iteration      6/ 20000 | consumed samples:        4,608 | consumed tokens:       75,497,472 | batch size: 768 | step time: 11361.95 ms | throughput: 458.12 tflop/s (model) | 472.05 tflop/s (hardware) | 8652.04 tokens/s/gpu | Memory allocated 8,692.08 MiB | max allocated 46,207.17 MiB | reserved 50,212.00 MiB | max reserved 50,212.00 MiB | global max reserved 50,212.00 MiB | learning rate: 3.000e-06 | loss scale:     1 | grad norm: nan | skipped iterations:   0 | nan iterations:   0 | average step time 14317.22 ms | remaining 3 days, 7:30:58  | completion 2025-04-05 00:55:25 (0.03 %) | language model loss: nan | run: 5
```

# 🎯 Expected Behavior

No NaNs during training

# 📜 Environment Information
Running on H100 GPUs with [sha-8866252](https://github.com/orgs/ServiceNow/packages/container/fast-llm/386329684?tag=sha-8866252)

# 📝 Additional Context

It is common practice for us to resume training from a linear merge of various checkpoints e.g., merging pretrained checkpoints from various iterations/epochs followed by SFT starting with the merged weights. We've seen 0 RMSNorm weights a few times already with the merged checkpoints, which prohibits us from using Fast-LLM directly on the merged checkpoint. The torch implementation of RMSNorm does not give the same error, and the gradients from huggingface's implementation do not show NaNs. 
The temporary workaround for us is to replace 0 RMSNorm weights with a small epsilon

**Comment by jlamypoirier** [comment]
We could fix this, but zero normalization weights mean ignoring entire sets of layer outputs so seems like a sign that something is going really wrong. Can you please elaborate on why this happens?

**Comment by sohamparikh** [comment]
This happens when at least one of the rmsnorm weights is 0. So far we've only seen this happen in the first transformer layer for only the `input_layernorm` (`norm_1` in Fast-LLM) after merging. We've only seen 1 or 2 weights (out of 5120) whose sum across individual checkpoints somehow turns out to be 0, so most (>99%) normalization weights are non-zero.

**Comment by jlamypoirier** [comment]
Did some digging, this comes from the memory-efficient implementation of layernorm/rmsnorm. In https://github.com/NVIDIA/apex/pull/1715 They deal with this problem by clamping the denominator by magnitude https://github.com/NVIDIA/apex/pull/1715/files#diff-c07d65ecf2d4a887113ba695f62be8e13ab9b3771d6bf6b2999f88959f25d3b0R496, which seems like a good idea. I'll add it later today.
